### PR TITLE
For Each Mediator

### PR DESCRIPTION
**Pull request by maheeka**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/synapse/config/xml/EnrichMediatorFactory.java [hunk 1]
```diff
@@ -96,7 +96,8 @@ private void populateSource(Source source, OMElement sourceEle) {
             OMAttribute xpathAttr = sourceEle.getAttribute(ATT_XPATH);
             if (xpathAttr != null && xpathAttr.getAttributeValue() != null) {
                 try {
-                    source.setXpath(SynapseXPathFactory.getSynapseXPath(sourceEle, ATT_XPATH));
+                    //source.setXpath(SynapseXPathFactory.getSynapseXPath(sourceEle, ATT_XPATH));
+                	source.setXpath(SynapsePathFactory.getSynapsePath(sourceEle, ATT_XPATH));
                 } catch (JaxenException e) {
                     handleException("Invalid XPath expression: " + xpathAttr);
```
Comment (isururanawaka) [inline modules/core/src/main/java/org/apache/synapse/config/xml/EnrichMediatorFactory.java]: Why exception caught is not passed to handleException method.

File: modules/core/src/main/java/org/apache/synapse/mediators/eip/EIPUtils.java [hunk 2]
```diff
@@ -39,7 +55,15 @@
...
 
     private static final Log log = LogFactory.getLog(EIPUtils.class);
-
+    
+    private static SynapseJsonPath rootJsonPath=null;
+    static{
+    	try{
+    		rootJsonPath=new SynapseJsonPath("$.");
+    	}catch(JaxenException e){
+    	}
```
Comment (isururanawaka) [inline modules/core/src/main/java/org/apache/synapse/mediators/eip/EIPUtils.java]: Do we need to ignore this exception ?

File: modules/core/src/main/java/org/apache/synapse/mediators/builtin/ForEachMediator.java [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+
+public class ForEachMediator extends AbstractMediator {
+
+	/** The jsonpath or xpath that will list the elements to be splitted */
+	private SynapsePath expression = null;
+
+	/** The target for the newly splitted messages */
+	private Target target = null;
+
+	public boolean mediate(MessageContext synCtx) {
```
Comment (kasun04) [inline modules/core/src/main/java/org/apache/synapse/mediators/builtin/ForEachMediator.java]: mediate(..) method is way too long...

File: modules/core/src/main/java/org/apache/synapse/mediators/builtin/ForEachMediator.java [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+
+		if (newEnvelope.getBody() != null) {
+
+			if (newEnvelope.getBody().getFirstElement() != null) {
+				newEnvelope.getBody().getFirstElement().detach();
+			}
+			newEnvelope.getBody().addChild(o);
+		}
+
+		newCtx.setEnvelope(newEnvelope);
```
Comment (kasun04) [inline modules/core/src/main/java/org/apache/synapse/mediators/builtin/ForEachMediator.java]: Why so many newlines????

File: modules/core/src/main/java/org/apache/synapse/mediators/eip/EIPUtils.java [hunk 5]
```diff
@@ -148,6 +172,47 @@ public static void enrichEnvelope(SOAPEnvelope envelope, SOAPEnvelope enricher,
...
+    	return getRootJSON(messageContext);
+    }
+    
+    public static Object getRootJSONObject(String json) throws JaxenException{
+    	return getRootJSON(json);
+    }
+    
+    private static Object getRootJSON(Object object) throws JaxenException{
+    	Object root=null;
+    	Object objectList=rootJsonPath.evaluate(object);
```
Comment (kasun04) [inline modules/core/src/main/java/org/apache/synapse/mediators/eip/EIPUtils.java]: Code formatting issues... have you configured your IDE properly????

File: modules/core/src/main/java/org/apache/synapse/mediators/eip/splitter/IterateMediator.java [hunk 6]
```diff
@@ -99,55 +103,77 @@ public boolean mediate(MessageContext synCtx) {
...
-                    synLog.traceOrDebug(
-                            "Submitting " + (msgNumber+1) + " of " + msgNumber +
-                            (target.isAsynchronous() ? " messages for processing in parallel" :
-                             " messages for processing in sequentially"));
+    
+                int msgCount = splitElements.size();
+                int msgNumber = 0;
+                
+                // iterate through the list
+                for (Object o : splitElements) {
```
Comment (kasun04) [inline modules/core/src/main/java/org/apache/synapse/mediators/eip/splitter/IterateMediator.java]: Use better naming...

File: modules/commons/src/main/java/org/apache/synapse/commons/json/JSONProviderUtil.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
+package org.apache.synapse.commons.json;
+
+import org.codehaus.jackson.map.ObjectMapper;
+import org.codehaus.jackson.map.ObjectWriter;
+/**
+ * This class contains some util methods to use with JSON providers which are used in JSONPath library
+ * 
+ */
+public class JSONProviderUtil {
```
Comment (kasun04) [inline modules/commons/src/main/java/org/apache/synapse/commons/json/JSONProviderUtil.java]: Util class with one method?

File: modules/core/src/main/java/org/apache/synapse/mediators/builtin/ForEachMediator.java [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+	 */
+	private Target target = null;
+
+	public boolean mediate(MessageContext synCtx) {
+		SynapseLog synLog = getLog(synCtx);
+
+		if (synLog.isTraceOrDebugEnabled()) {
+			synLog.traceOrDebug("Start : Foreach mediator");
+			// synLog.traceOrDebug("FE*=Message : " + synCtx.getEnvelope());
+			// synLog.traceTrace("FE*=Message : " + synCtx.getEnvelope());
```
Comment (vanjikumaran) [inline modules/core/src/main/java/org/apache/synapse/mediators/builtin/ForEachMediator.java]: Remove unwanted code lines

File: modules/core/src/main/java/org/apache/synapse/mediators/builtin/ForEachMediator.java (hunk 4 above)
Comment (maheeka) [inline modules/core/src/main/java/org/apache/synapse/mediators/builtin/ForEachMediator.java]: @kasun04 This is a new class for ForEachMediator

File: modules/core/src/main/java/org/apache/synapse/mediators/eip/EIPUtils.java (hunk 5 above)
Comment (maheeka) [inline modules/core/src/main/java/org/apache/synapse/mediators/eip/EIPUtils.java]: @kasun04 Resolved.

File: modules/core/src/main/java/org/apache/synapse/config/xml/EnrichMediatorFactory.java (hunk 1 above)
Comment (maheeka) [inline modules/core/src/main/java/org/apache/synapse/config/xml/EnrichMediatorFactory.java]: Ignored (Pulled from now deprecated json implementation)

File: modules/core/src/main/java/org/apache/synapse/mediators/eip/EIPUtils.java (hunk 2 above)
Comment (maheeka) [inline modules/core/src/main/java/org/apache/synapse/mediators/eip/EIPUtils.java]: Ignored (Pulled from now deprecated json implementation)

File: modules/core/src/main/java/org/apache/synapse/mediators/builtin/ForEachMediator.java (hunk 3 above)
Comment (maheeka) [inline modules/core/src/main/java/org/apache/synapse/mediators/builtin/ForEachMediator.java]: @kasun04 Resolved with removal of json implementation

File: modules/commons/src/main/java/org/apache/synapse/commons/json/JSONProviderUtil.java (hunk 7 above)
Comment (maheeka) [inline modules/commons/src/main/java/org/apache/synapse/commons/json/JSONProviderUtil.java]: @kasun04 Removed. This was pulled from now removed json implementation.

File: modules/core/src/main/java/org/apache/synapse/mediators/builtin/ForEachMediator.java (hunk 8 above)
Comment (maheeka) [inline modules/core/src/main/java/org/apache/synapse/mediators/builtin/ForEachMediator.java]: Resolved

File: modules/core/src/main/java/org/apache/synapse/mediators/eip/splitter/IterateMediator.java (hunk 6 above)
Comment (maheeka) [inline modules/core/src/main/java/org/apache/synapse/mediators/eip/splitter/IterateMediator.java]: Resolved in ForEachMediator.java

**Pull request closed without merging**